File: Submissions/486_Predict_the_Winner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        return recursive(nums, 0, nums.size()-1) >= 0;
    }
    
    int recursive(vector<int>& nums, int begin, int end){
        if(begin==end) return nums[begin];
        
        return std::max(nums[begin] - recursive(nums, begin+1, end), nums[end] - recursive(nums, begin, end-1));    
    }
};

File: Kotlin/486_Predict_the_Winner.kt
class Solution {
    fun PredictTheWinner(nums: IntArray): Boolean {
        return recursive(nums, 0, nums.lastIndex) >= 0
    }
    
    fun recursive(nums: IntArray, begin: Int, end: Int ): Int{
        if(begin==end) return nums[begin]
        return maxOf(nums[begin] - recursive(nums, begin+1, end), nums[end] - recursive(nums, begin, end-1))
    }
}